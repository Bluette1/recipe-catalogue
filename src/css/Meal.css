@import url('https://fonts.googleapis.com/css2?family=Mountains+of+Christmas&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Elsie&family=Mountains+of+Christmas&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Elsie&family=Milonga&family=Mountains+of+Christmas&display=swap');

.meal-row {
  height: 200px;
  padding: 32px 119px 26px 27px;
  margin-top: 40px;
  display: flex;
  justify-content: space-evenly;
  margin-right: 100px;
  margin-left: 100px;
  border-radius: 4px;
  border: solid 1px #e8e8e8;
  background-color: #d57d1f;
}

.category {
  width: 49px;
  height: 18px;
  margin: 0 39px 0 0;
  opacity: 0.5;
  font-family: 'Milonga', cursive;
  font-size: 14px;
  font-weight: bold;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: normal;
  color: #121212;
}

.title {
  width: 197px;
  height: 58px;
  margin: 3px 323px 0 0;
  font-family: 'Elsie', cursive;
  font-size: 22px;
  font-weight: bold;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: -0.2px;
  color: #121212;
}

.title-category,
.image-area .right {
  width: 250px;
}

.image-area {
  height: 150px;
  width: 150px;
  border: solid 3px #2d2013;
}

a {
  text-decoration: none;
}

.tags {
  margin-left: 53px;
  opacity: 0.5;
  font-family: 'Milonga', cursive;
  font-size: 14px;
  font-weight: bold;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: normal;
  color: #121212;
}

.meal-row button {
  font-family: 'Elsie', cursive;
  font-size: 13px;
  font-weight: 300;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: 0.5px;
  text-align: center;
  color: #fff;
  width: 160px;
  height: 33px;
  margin: 12px 0 9px 59px;
  border-radius: 3px;
  background-color: #2d2013;
  border: #2d2013;
  cursor: pointer;
  padding: 1px;
  border-right: 1px solid #8b8b8b;
  border-left: 1px solid #fff;
  border-bottom: 1px solid #fff;
}

.meal-row button:hover {
  border: 2px outset #def;
  color: #121212;
  background-color: #f8f8f8 !important;
}

.hidden {
  visibility: hidden;
}

.highlighted {
  background-color: yellow;
} 

@media (max-width: 1250px) {
  .meal-row {
    display: block;
    height: fit-content;
    margin-left: 0;
    margin-right: 0;
  }

  .meal-row  button, .tags {
    margin: 10px 10px 0 0;
  }
}
