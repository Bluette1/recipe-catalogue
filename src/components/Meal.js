import React from 'react';
import PropTypes from 'prop-types';
import '../css/Meal.css';
import cx from 'classnames';

const Meal = ({
  meal, highlightMeal, hideFromList,
}) => {
  const {
    strMeal, strCategory, strMealThumb, strTags, strArea, strYoutube, strInstructions, idMeal,
  } = meal;

  const parseNum = str => {
    const { length } = str;
    let index = 1;
    while (Number.isFinite(parseInt(str.charAt(length - index), 10))) {
      index += 1;
    }
    return str.slice(-1 * (index - 1));
  };

  const ingredients = () => {
    let ingredientsStr = '';
    const keys = Object.keys(meal);
    for (let idx = 0; idx < keys.length; idx += 1) {
      if (keys[idx].includes('strIngredient') && meal[keys[idx]] !== null && meal[keys[idx]] !== '') {
        const num = parseNum(keys[idx]);
        ingredientsStr += `${meal[keys[idx]]}: ${meal[`strMeasure${num}`]}, `;
      }
    }
    const resultIngredientsStr = ingredientsStr.slice(0, -2);
    return resultIngredientsStr;
  };

  const detailsUrl = () => `/details?img=${strMealThumb}&t=${strMeal}
  &i=${ingredients()}&y=${strYoutube}&ins=${strInstructions}`;

  return (
    <div
      className={cx(
        'meal-row',
        meal.highlight && 'highlighted',
        meal.hide && 'hidden',
      )}
    >
      <div className="title-category">
        <p className="category">{strCategory}</p>
        <p className="area category">{strArea}</p>
        <h4 className="title">{strMeal}</h4>
      </div>
      <div
        className="image-area"
        style={{
          backgroundImage: `url(${strMealThumb})`,
          backgroundRepeat: 'no-repeat',
          backgroundPosition: '50% 50%',
          backgroundSize: 'cover',
        }}
      />
      <div className="right">
        <h4 className="tags">
          Tags:&nbsp;
          {strTags}
        </h4>
        <a href={detailsUrl()}><button type="button" className="details">View details</button></a>
        <button type="button" className="hide" onClick={() => hideFromList(idMeal)}>Hide from list</button>
        <button type="button" className="highlight" onClick={() => highlightMeal(idMeal)}>Highlight meal</button>
      </div>

    </div>
  );
};

Meal.propTypes = {
  meal: PropTypes.objectOf(PropTypes.string).isRequired,
  highlightMeal: PropTypes.func.isRequired,
  hideFromList: PropTypes.func.isRequired,
};

export default Meal;
