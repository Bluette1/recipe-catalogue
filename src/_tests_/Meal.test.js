import React from 'react';
import {
  render, cleanup, fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore, { history } from '../store';
import Meal from '../components/Meal';

configure({ adapter: new Adapter() });

afterEach(cleanup);

const store = configureStore();
const meal = {
  idMeal: '1',
  strMeal: 'Mealia',
  strCategory: 'Dessert',
  strMealThumb: 'image-url',
  strTags: 'dessert, British',
  strArea: 'British',
  strYoutube: 'Youtube url',
  strInstructions: 'How to Cook Mealia',
  strIngredient1: 'corn',
  strMeasure1: '1 cup',
};

const hideFromListSpy = jest.fn();
const highLightMealSpy = jest.fn();

const MealWithStore = () => (
  <Provider store={store}>
    <React.StrictMode>
      <ConnectedRouter history={history}>
        <Meal hideFromList={hideFromListSpy} highlightMeal={highLightMealSpy} meal={meal} />
      </ConnectedRouter>
    </React.StrictMode>
  </Provider>
);

it('renders the Meal component', () => {
  const rendered = render(<MealWithStore />);
  const mealTitle = rendered.getByText('Mealia');
  expect(mealTitle).toBeInTheDocument();

  const mealCategory = rendered.getByText('Dessert');
  expect(mealCategory).toBeInTheDocument();

  const streetArea = rendered.getByText('British');
  expect(streetArea).toBeInTheDocument();

  const tags = rendered.getByText('Tags: dessert, British');
  expect(tags).toBeInTheDocument();
  expect(rendered).toMatchSnapshot();
});
it('renders the Meal component with the correct buttons', () => {
  const rendered = mount(<MealWithStore />);
  expect(rendered.find('button.details').text()).toBe('View details');
  expect(rendered.find('button.hide').text()).toBe('Hide from list');
  expect(rendered).toMatchSnapshot();
});
it('renders the Meal component with the correct link to the meal details page', () => {
  const rendered = mount(<MealWithStore />);
  const link = rendered.find('a').get(0);

  const detailsUrl = `/details?img=image-url&t=Mealia&c=Dessert
  &i=corn: 1 cup&st=British&y=Youtube url&ins=How to Cook Mealia`;
  expect(link.props.href).toBe(detailsUrl);
  expect(rendered).toMatchSnapshot();
});

it('renders the Meal component with the `hide meal` button working correctly', () => {
  const rendered = render(<MealWithStore />);
  const hideButton = rendered.getByText('Hide from list');
  expect(hideButton).toBeInTheDocument();
  fireEvent.click(hideButton);
  expect(hideFromListSpy).toHaveBeenCalledWith(meal.idMeal);
  expect(rendered).toMatchSnapshot();
});

it('renders the Meal component with the `highlight meal` button working correctly', () => {
  const rendered = render(<MealWithStore />);
  const highlightButton = rendered.getByText('Highlight meal');
  expect(highlightButton).toBeInTheDocument();
  fireEvent.click(highlightButton);
  expect(highLightMealSpy).toHaveBeenCalledWith(meal.idMeal);
  expect(rendered).toMatchSnapshot();
});
