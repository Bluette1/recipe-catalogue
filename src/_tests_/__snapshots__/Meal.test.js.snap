// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Meal component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="meal-row"
      >
        <div
          class="title-category"
        >
          <p
            class="category"
          >
            Dessert
          </p>
          <p
            class="area category"
          >
            British
          </p>
          <h4
            class="title"
          >
            Mealia
          </h4>
        </div>
        <div
          class="image-area"
          style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
        />
        <div
          class="right"
        >
          <h4
            class="tags"
          >
            Tags: 
            dessert, British
          </h4>
          <a
            href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
          >
            <button
              class="details"
              type="button"
            >
              View details
            </button>
          </a>
          <button
            class="hide"
            type="button"
          >
            Hide from list
          </button>
          <button
            class="highlight"
            type="button"
          >
            Highlight meal
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="meal-row"
    >
      <div
        class="title-category"
      >
        <p
          class="category"
        >
          Dessert
        </p>
        <p
          class="area category"
        >
          British
        </p>
        <h4
          class="title"
        >
          Mealia
        </h4>
      </div>
      <div
        class="image-area"
        style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
      />
      <div
        class="right"
      >
        <h4
          class="tags"
        >
          Tags: 
          dessert, British
        </h4>
        <a
          href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
        >
          <button
            class="details"
            type="button"
          >
            View details
          </button>
        </a>
        <button
          class="hide"
          type="button"
        >
          Hide from list
        </button>
        <button
          class="highlight"
          type="button"
        >
          Highlight meal
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders the Meal component with the \`hide meal\` button working correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="meal-row"
      >
        <div
          class="title-category"
        >
          <p
            class="category"
          >
            Dessert
          </p>
          <p
            class="area category"
          >
            British
          </p>
          <h4
            class="title"
          >
            Mealia
          </h4>
        </div>
        <div
          class="image-area"
          style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
        />
        <div
          class="right"
        >
          <h4
            class="tags"
          >
            Tags: 
            dessert, British
          </h4>
          <a
            href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
          >
            <button
              class="details"
              type="button"
            >
              View details
            </button>
          </a>
          <button
            class="hide"
            type="button"
          >
            Hide from list
          </button>
          <button
            class="highlight"
            type="button"
          >
            Highlight meal
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="meal-row"
    >
      <div
        class="title-category"
      >
        <p
          class="category"
        >
          Dessert
        </p>
        <p
          class="area category"
        >
          British
        </p>
        <h4
          class="title"
        >
          Mealia
        </h4>
      </div>
      <div
        class="image-area"
        style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
      />
      <div
        class="right"
      >
        <h4
          class="tags"
        >
          Tags: 
          dessert, British
        </h4>
        <a
          href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
        >
          <button
            class="details"
            type="button"
          >
            View details
          </button>
        </a>
        <button
          class="hide"
          type="button"
        >
          Hide from list
        </button>
        <button
          class="highlight"
          type="button"
        >
          Highlight meal
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders the Meal component with the \`highlight meal\` button working correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="meal-row"
      >
        <div
          class="title-category"
        >
          <p
            class="category"
          >
            Dessert
          </p>
          <p
            class="area category"
          >
            British
          </p>
          <h4
            class="title"
          >
            Mealia
          </h4>
        </div>
        <div
          class="image-area"
          style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
        />
        <div
          class="right"
        >
          <h4
            class="tags"
          >
            Tags: 
            dessert, British
          </h4>
          <a
            href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
          >
            <button
              class="details"
              type="button"
            >
              View details
            </button>
          </a>
          <button
            class="hide"
            type="button"
          >
            Hide from list
          </button>
          <button
            class="highlight"
            type="button"
          >
            Highlight meal
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="meal-row"
    >
      <div
        class="title-category"
      >
        <p
          class="category"
        >
          Dessert
        </p>
        <p
          class="area category"
        >
          British
        </p>
        <h4
          class="title"
        >
          Mealia
        </h4>
      </div>
      <div
        class="image-area"
        style="background-image: url(image-url); background-repeat: no-repeat; background-position: 50% 50%; background-size: cover;"
      />
      <div
        class="right"
      >
        <h4
          class="tags"
        >
          Tags: 
          dessert, British
        </h4>
        <a
          href="/details?img=image-url&t=Mealia
  &i=corn: 1 cup&y=Youtube url&ins=How to Cook Mealia"
        >
          <button
            class="details"
            type="button"
          >
            View details
          </button>
        </a>
        <button
          class="hide"
          type="button"
        >
          Hide from list
        </button>
        <button
          class="highlight"
          type="button"
        >
          Highlight meal
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders the Meal component with the correct buttons 1`] = `ReactWrapper {}`;

exports[`renders the Meal component with the correct link to the meal details page 1`] = `ReactWrapper {}`;
